mobilecoinofficial/mobilecoin
# MCC-1698 Verify IAS reports using the new framework

Soundtrack of this PR: [link to song that really fits the mood of this PR]()

### Motivation

This ports away from the old `VerificationReport::verify()` method to use the new verifiers framework.

### In this PR

 * `mc-attest-ake`:
    - Remove `failure` for `displaydoc`.
    - Remove `mc-sgx-build` for `mc_util_build_{script, sgx}`.
    - Add comment-wrapping rustfmt.toml.
    - Use `VerifierError` instead of `VerifyError`.
    - Split `AuthRequestInput<Handshake, ...>` input event into `ClientAuthRequestInput` (which does nothing extra) and `NodeAuthRequestInput` (which contains a verifier for the peer node's report).
    - Remove the expected measurements stuff from the states, in favor of verifiers attached by the caller to input events.
    - Fix object safety issue on Mealy machine `Transition` trait/implementations.
    - Return the parsed `VerificationReportData`, instead of the fake `AuthSuccess` type (verifier failures already contain this data).
    - 
 * `mc_attest_core`:
    - `Verifier` now stores trust anchors as DER byte vectors, not `Certificate`s, which aren't `Send`/`Sync` safe.
    - Update `MrSignerVerifier`, `MrEnclaveVerifier` to use the same non-consuming pattern as `Verifier`.
    - Add `From<mc_sgx_css::Signature>` implementations to `MrSignerVerifier`, `MrEnclaveVerifier`
    - Add `From<[u8; 32]>` implementations for `MrSigner` and `MrEnclave`.
 * `mc_sgx_css`:
    - Remove `failure` for `displaydoc` (also fixes improper/unnoticed `std` linkage).
 * `mc_attest_enclave_api`:
    - Remove `failure` for `displaydoc`.
    - Change use of `VerifyError` to `VerifierError`.
    - Remove the handshake pattern from the `Session` trait (no longer used).
 * `mc_crypto_ake_enclave`:
    - Add a `get_verifier()` private method to create a proper verifier for the current enclave.
    - Use it wherever we need to verify a report.
    - Copy-pasta the old contents of `accept()` into `peer_accept()` and `client_accept()` to account for the injection of the verifier via input event.
 * `mc-ledger-sync`, `mc-mobilecoind`:
    - Use a verifier constructed from the `sigstruct()` instead of building an expected measurements array.
 * `mc-sgx-report-cache-untrusted`:
    - Adapt to the use of `VerifierError(report_data)` instead of the deep `VerifyError` hierarchy when trying to handle a platform info blob.

### Future Work
* Fix broken unit tests.
* Deprecate and remove old `VerificationReport::verify()` methods.
* Verifier/V4 API support for the new-style  debaidufied crates
* Constant time verification
